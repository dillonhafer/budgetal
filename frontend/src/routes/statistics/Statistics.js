import React, { Component } from 'react';
import { FindStatisticRequest } from '@shared/api/statistics';
import { title, scrollTop, error } from 'window';
import { availableYears, currencyf, monthName } from '@shared/helpers';
import { Text, Heading, Select, Pane } from 'evergreen-ui';
import times from 'lodash/times';
import './statistics.css';
import StatisticsChart from './StatisticsChart';
import Spinner from 'components/Spinner';

class Statistics extends Component {
  state = {
    loading: false,
    budgetCategories: [],
  };

  componentDidMount() {
    const { month, year } = this.props.match.params;
    title(`${monthName(month)} ${year} | Statistics`);
    scrollTop();
    this.loadStatistics();
  }

  missing() {
    return (
      <Pane textAlign="center" width="100%" margin="80px">
        <h2>Uh-oh!</h2>
        <p>It looks like you don't have a budget for this month</p>
      </Pane>
    );
  }

  handleOnDateChange = e => {
    const date = {
      year: this.props.match.params.year,
      month: this.props.match.params.month,
      [e.target.name]: e.target.value,
    };

    this.props.history.push(`/monthly-statistics/${date.year}/${date.month}`);
  };

  loadStatistics = async () => {
    this.setState({ loading: true });
    try {
      const resp = await FindStatisticRequest(this.props.match.params);
      if (resp && resp.ok) {
        const totalSpent = resp.budgetCategories.reduce(
          (acc, cat) => acc + parseFloat(cat.amountSpent),
          0.0,
        );
        const budgetCategories = resp.budgetCategories.map(cat => {
          const name = cat.name;
          const amountSpent = parseFloat(cat.amountSpent);

          let percentSpent = 0;
          if (parseFloat(cat.amountSpent) > 0) {
            percentSpent = Math.round(
              (parseFloat(cat.amountSpent) / totalSpent) * 100,
            );
          }
          return {
            name,
            amountSpent,
            percentSpent,
          };
        });
        this.setState({ budgetCategories });
      }
    } catch (err) {
      error(err);
    } finally {
      this.setState({ loading: false });
    }
  };

  render() {
    const { loading } = this.state;
    const { month, year } = this.props.match.params;
    return (
      <Pane paddingLeft={24} paddingRight={24}>
        <Pane
          display="flex"
          flexDirection="row"
          alignItems="center"
          justifyContent="space-between"
          marginBottom={16}
        >
          <Heading size={800}>
            STATISTICS FOR {monthName(month).toUpperCase()} {year}
          </Heading>
          <Pane>
            <Select
              defaultValue={month}
              name="month"
              onChange={this.handleOnDateChange}
              flex="unset"
              width={90}
              marginRight={4}
            >
              {times(12, i => (
                <option key={i} value={i + 1}>
                  {monthName(i + 1)}
                </option>
              ))}
            </Select>
            <Select
              defaultValue={year}
              name="year"
              onChange={this.handleOnDateChange}
              flex="unset"
              width={70}
            >
              {availableYears().map(y => {
                return (
                  <option key={y} value={y.toString()}>
                    {y}
                  </option>
                );
              })}
            </Select>
          </Pane>
        </Pane>

        <Spinner visible={loading} />
        <Pane display={'flex'} flexDirection="row">
          {loading || this.state.budgetCategories.length > 0
            ? null
            : this.missing()}
          <Pane display="flex" flex="1">
            <StatisticsChart budgetCategories={this.state.budgetCategories} />
          </Pane>
          <Pane display="flex" flex="1">
            <ul className="stat-list" style={{ width: '100%' }}>
              {this.state.budgetCategories.map((category, key) => {
                const statIconClass =
                  'stat-icon stat-icon-' +
                  category.name.toLowerCase().replace('/', '-');
                return (
                  <li key={key}>
                    <div className="stat-list-item">
                      <Text>
                        <div className={statIconClass} />
                        <b>{category.name}</b>
                        <br />
                        <span className="percentSpent">
                          {currencyf(category.amountSpent)} -{' '}
                          {category.percentSpent}%
                        </span>
                      </Text>
                    </div>
                  </li>
                );
              })}
            </ul>
          </Pane>
        </Pane>
      </Pane>
    );
  }
}

export default Statistics;
