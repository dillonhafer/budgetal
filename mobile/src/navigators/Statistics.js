import React from 'react';
import { View } from 'react-native';
import { StackNavigator } from 'react-navigation';

import TabletNavigator from './TabletNavigator';

// Screens
import StatisticsScreen from 'screens/statistics/Statistics';
import MonthlyChartScreen from 'screens/statistics/MonthlyChart';

import { Ionicons } from '@expo/vector-icons';

const headerStyle = {
  height: 44,
};
const sidebarHeaderStyle = {
  height: 43.5,
};

const AccountNavigatorStack = StackNavigator(
  {
    Statistics: {
      screen: StatisticsScreen,
      navigationOptions: { headerStyle },
    },
  },
  {
    cardStyle: {
      backgroundColor: '#ececec',
      shadowOpacity: 0,
    },
  },
);

const StatisticsSidebarNavigatorStack = StackNavigator(
  {
    MonthlyChart: {
      screen: MonthlyChartScreen,
      navigationOptions: ({ navigation }) => {
        return {
          headerStyle: sidebarHeaderStyle,
          title: `Chart View`,
        };
      },
    },
  },
  {},
);

class StatisticsNavigator extends TabletNavigator {
  MainNavigator = AccountNavigatorStack;
  SideNavigator = StatisticsSidebarNavigatorStack;

  static navigationOptions = {
    header: null,
    tabBarLabel: 'Statistics',
    tabBarIcon: ({ tintColor }) => (
      <Ionicons name="md-stats" size={32} color={tintColor} />
    ),
  };
}

export default StatisticsNavigator;
