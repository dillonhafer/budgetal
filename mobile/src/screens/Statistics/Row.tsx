import { currencyf } from "@shared/helpers";
import CategoryImage from "@src/components/CategoryImage";
import { Bold } from "@src/components/Text";
import React from "react";
import styled from "styled-components/native";
import { GetMonthlyStatistics_monthlyStatistic } from "./__generated__/GetMonthlyStatistics";

export interface Category extends GetMonthlyStatistics_monthlyStatistic {}

const rowPadding = 20;

const CurrencyRow = styled.View({
  marginLeft: rowPadding,
  flex: 1,
  justifyContent: "center",
});

const BoldText = styled(Bold)({
  color: "#444",
  fontSize: 18,
  marginBottom: 5,
});

const Container = styled.View({
  backgroundColor: "white",
  padding: rowPadding,
  justifyContent: "center",
  flexDirection: "row",
});

const Row = ({ category }: { category: Category }) => (
  <Container>
    <CategoryImage name={category.name} />
    <CurrencyRow>
      <BoldText>{category.name}</BoldText>
      <BoldText>
        {currencyf(category.amountSpent)} - {category.percentSpent}%
      </BoldText>
    </CurrencyRow>
  </Container>
);

export default Row;
