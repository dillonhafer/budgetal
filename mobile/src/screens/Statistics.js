import React, { Component } from 'react';
import {
  StyleSheet,
  Image,
  Text,
  StatusBar,
  FlatList,
  View,
} from 'react-native';

// Components
import { Ionicons } from '@expo/vector-icons';
import { FindStatisticRequest } from 'api/statistics';
import { currencyf, categoryImage } from 'utils/helpers';

class StatisticsScreen extends Component {
  static navigationOptions = {
    title: 'Statistics',
    tabBarIcon: ({ tintColor }) => (
      <Ionicons name="md-stats" size={32} color={tintColor} />
    ),
  };

  state = {
    loading: false,
    budgetCategories: [],
  };

  componentDidMount() {
    this.loadStatistics();
  }

  renderCategory = ({ item: budgetCategory }) => {
    const percentSpent = 37;
    return (
      <View style={styles.categoryRow}>
        <View style={{ flexDirection: 'row' }}>
          <Image
            style={styles.categoryImage}
            source={categoryImage(budgetCategory.name)}
          />
          <View
            style={{
              flexDirection: 'column',
              flex: 1,
              justifyContent: 'center',
            }}
          >
            <Text style={styles.categoryName}>{budgetCategory.name}</Text>
            <Text style={styles.categoryName}>
              {currencyf(budgetCategory.amountSpent)} -{' '}
              {budgetCategory.percentSpent}%
            </Text>
          </View>
        </View>
      </View>
    );
  };

  loadStatistics = async () => {
    this.setState({ loading: true });
    try {
      const resp = await FindStatisticRequest({ month: 12, year: 2017 });
      if (resp && resp.ok) {
        const totalSpent = resp.budgetCategories.reduce(
          (acc, cat) => acc + parseFloat(cat.amountSpent),
          0.0,
        );
        const budgetCategories = resp.budgetCategories.map(cat => {
          const name = cat.name;
          const amountSpent = parseFloat(cat.amountSpent);

          let percentSpent = 0;
          if (parseFloat(cat.amountSpent) > 0) {
            percentSpent = Math.round(
              parseFloat(cat.amountSpent) / totalSpent * 100,
            );
          }
          return {
            key: cat.name,
            name,
            amountSpent,
            percentSpent,
          };
        });
        this.setState({ budgetCategories });
      }
    } catch (err) {
      // error(err);
    } finally {
      this.setState({ loading: false });
    }
  };

  renderSeparator = () => {
    return (
      <View
        style={{
          height: 1,
          width: '100%',
          backgroundColor: '#CED0CE',
        }}
      />
    );
  };

  render() {
    const { loading, budgetCategories } = this.state;
    return (
      <View style={styles.container}>
        <StatusBar barStyle="dark-content" />
        <FlatList
          style={styles.list}
          refreshing={loading}
          onRefresh={this.loadStatistics}
          data={budgetCategories}
          ItemSeparatorComponent={this.renderSeparator}
          renderItem={this.renderCategory}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    flexDirection: 'column',
  },
  list: {
    alignSelf: 'stretch',
  },
  categoryRow: {
    backgroundColor: '#fff',
    padding: 15,
    justifyContent: 'center',
  },
  categoryName: {
    fontWeight: '700',
    color: '#444',
    fontSize: 18,
    marginBottom: 5,
  },
  categoryImage: {
    width: 64,
    height: 64,
    margin: 10,
  },
});

export default StatisticsScreen;
