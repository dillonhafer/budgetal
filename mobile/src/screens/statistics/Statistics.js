import React, { PureComponent } from 'react';
import {
  StyleSheet,
  Image,
  Text,
  StatusBar,
  FlatList,
  View,
  RefreshControl,
} from 'react-native';

// Components
import { Ionicons } from '@expo/vector-icons';
import { FindStatisticRequest } from 'api/statistics';
import { currencyf, categoryImage } from 'utils/helpers';
import DatePicker from 'utils/DatePicker';
import Spin from 'utils/Spin';

class StatisticsScreen extends PureComponent {
  static navigationOptions = {
    title: 'Statistics',
    tabBarIcon: ({ tintColor }) => (
      <Ionicons name="md-stats" size={32} color={tintColor} />
    ),
  };

  state = {
    loading: false,
    refreshing: false,
    budget: {
      month: new Date().getMonth() + 1,
      year: new Date().getFullYear(),
    },
    budgetCategories: [],
  };

  componentDidMount() {
    setTimeout(() => {
      this.loadStatistics({
        month: new Date().getMonth() + 1,
        year: new Date().getFullYear(),
      });
    }, 800);
  }

  renderCategory = ({ item: budgetCategory }) => {
    return (
      <View style={styles.categoryRow}>
        <View style={{ flexDirection: 'row' }}>
          <Image
            style={styles.categoryImage}
            source={categoryImage(budgetCategory.name)}
          />
          <View
            style={{
              flexDirection: 'column',
              flex: 1,
              justifyContent: 'center',
            }}
          >
            <Text style={styles.categoryName}>{budgetCategory.name}</Text>
            <Text style={styles.categoryName}>
              {currencyf(budgetCategory.amountSpent)} -{' '}
              {budgetCategory.percentSpent}%
            </Text>
          </View>
        </View>
      </View>
    );
  };

  loadStatistics = async ({ month, year }) => {
    this.setState({ loading: true });
    try {
      const resp = await FindStatisticRequest({ month, year });
      if (resp && resp.ok) {
        const totalSpent = resp.budgetCategories.reduce(
          (acc, cat) => acc + parseFloat(cat.amountSpent),
          0.0,
        );
        const budgetCategories = resp.budgetCategories.map(cat => {
          const name = cat.name;
          const amountSpent = parseFloat(cat.amountSpent);

          let percentSpent = 0;
          if (parseFloat(cat.amountSpent) > 0) {
            percentSpent = Math.round(
              parseFloat(cat.amountSpent) / totalSpent * 100,
            );
          }
          return {
            key: cat.name,
            name,
            amountSpent,
            percentSpent,
          };
        });
        this.setState({ budgetCategories });
      }
    } catch (err) {
      console.log(err);
    } finally {
      this.setState({ loading: false });
    }
  };

  renderSeparator = () => {
    return (
      <View
        style={{
          height: 1,
          width: '100%',
          backgroundColor: '#CED0CE',
        }}
      />
    );
  };

  onDateChange = ({ month, year }) => {
    if (
      String(month) !== String(this.state.budget.month) ||
      String(year) !== String(this.state.budget.year)
    ) {
      this.setState({ budget: { month, year } });
      this.loadStatistics({ month, year });
    }
  };

  onRefresh = async () => {
    this.setState({ refreshing: true });
    try {
      const { year, month } = this.state.budget;
      await this.loadStatistics({ year, month });
    } catch (err) {
      console.log(err);
    } finally {
      this.setState({ refreshing: false });
    }
  };

  render() {
    const { refreshing, loading, budget, budgetCategories } = this.state;
    return (
      <View style={styles.container}>
        <StatusBar barStyle="dark-content" />
        <DatePicker
          month={budget.month}
          year={budget.year}
          onChange={this.onDateChange}
        />
        <FlatList
          style={styles.list}
          refreshControl={
            <RefreshControl
              tintColor={'lightskyblue'}
              refreshing={refreshing}
              onRefresh={this.onRefresh}
            />
          }
          data={budgetCategories}
          ItemSeparatorComponent={this.renderSeparator}
          renderItem={this.renderCategory}
        />
        <Spin spinning={loading && !refreshing} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    flexDirection: 'column',
  },
  list: {
    alignSelf: 'stretch',
  },
  categoryRow: {
    backgroundColor: '#fff',
    padding: 15,
    justifyContent: 'center',
  },
  categoryName: {
    fontWeight: '700',
    color: '#444',
    fontSize: 18,
    marginBottom: 5,
  },
  categoryImage: {
    width: 64,
    height: 64,
    margin: 10,
  },
});

export default StatisticsScreen;
