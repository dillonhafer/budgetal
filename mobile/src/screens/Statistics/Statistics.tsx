import { useQuery } from "react-apollo";
import { FontAwesome } from "@expo/vector-icons";
import { colors } from "@shared/theme";
import { ListSeparator } from "@src/components/ListRow";
import { Bold } from "@src/components/Text";
import DatePicker from "@src/utils/DatePicker";
import { BlurViewInsetProps } from "@src/utils/navigation-helpers";
import Spin from "@src/utils/Spin";
import gql from "graphql-tag";
import React, { useEffect, useState } from "react";
import { FlatList, RefreshControl, StatusBar } from "react-native";
import styled from "styled-components/native";
import Row, { Category } from "./Row";

const GET_MONTHLY_STATISTICS = gql`
  query GetMonthlyStatistics($year: Int!, $month: Int!) {
    monthlyStatistic(year: $year, month: $month) {
      amountSpent
      totalSpent
      percentSpent
      id
      month
      name
      year
    }
  }
`;

const MissingContainer = styled.View({
  padding: 20,
  paddingTop: 40,
  alignItems: "center",
});

const MissingText = styled(Bold)({
  margin: 5,
  textAlign: "center",
});

const defaultYear = parseInt(`${new Date().getFullYear()}`);
const defaultMonth = parseInt(`${new Date().getMonth() + 1}`);

interface PickerProps {
  year: number;
  month: number;
}

const Container = styled.View({
  flex: 1,
  backgroundColor: "#fff",
});

const Empty = () => (
  <MissingContainer>
    <FontAwesome name="money" size={32} color={colors.success} />
    <MissingText>There aren't any items yet</MissingText>
  </MissingContainer>
);

const StatisticsScreen = () => {
  const [year, setYear] = useState(defaultYear);
  const [month, setMonth] = useState(defaultMonth);
  const [refreshing, setRefreshing] = useState(false);
  const { loading, data, refetch } = useQuery(GET_MONTHLY_STATISTICS, {
    variables: { year, month },
  });
  const { monthlyStatistic = [] } = data;

  useEffect(() => {
    if (refreshing) {
      refetch({ year, month });
      setRefreshing(false);
    }
  }, [year, refreshing]);

  const onDateChange = ({ month, year }: PickerProps) => {
    setMonth(month);
    setYear(year);
  };

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <Container>
        <FlatList
          {...BlurViewInsetProps}
          contentInsetAdjustmentBehavior="automatic"
          ListHeaderComponent={
            <DatePicker year={year} month={month} onChange={onDateChange} />
          }
          ListEmptyComponent={<Empty />}
          refreshControl={
            <RefreshControl
              tintColor={"lightskyblue"}
              refreshing={refreshing}
              onRefresh={() => {
                setRefreshing(true);
              }}
            />
          }
          keyExtractor={(i: Category) => i.id}
          data={monthlyStatistic}
          ItemSeparatorComponent={ListSeparator}
          renderItem={({ item }: { item: Category }) => <Row category={item} />}
        />
        <Spin spinning={loading && !refreshing} />
      </Container>
    </>
  );
};

StatisticsScreen.navigationOptions = {
  title: "Statistics",
};

export default StatisticsScreen;
