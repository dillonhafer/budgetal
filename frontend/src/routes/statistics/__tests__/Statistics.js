import React from 'react';
import { shallow, mount } from 'enzyme';
import Statistics from '../Statistics';

const defaultProps = {
  match: {
    params: {
      year: 2018,
      month: 8,
    },
  },
};

describe('Statistics', () => {
  it('renders', () => {
    const comp = shallow(<Statistics {...defaultProps} />);
    expect(comp).toMatchSnapshot();
  });

  it('has the correct header', () => {
    const comp = mount(<Statistics {...defaultProps} />);
    expect(comp.find('Heading').text()).toEqual('STATISTICS FOR AUGUST 2018');
  });
});
