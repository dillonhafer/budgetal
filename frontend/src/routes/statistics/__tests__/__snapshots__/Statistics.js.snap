// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics renders 1`] = `
<withTheme(Pane)>
  <withTheme(Pane)
    alignItems="center"
    display="flex"
    flexDirection="row"
    justifyContent="space-between"
    marginBottom={16}
  >
    <withTheme(Heading)
      size={800}
    >
      STATISTICS FOR 
      AUGUST
       
      2018
    </withTheme(Heading)>
    <withTheme(Pane)>
      <withTheme(Select)
        defaultValue={8}
        flex="unset"
        marginRight={4}
        name="month"
        onChange={[Function]}
        width={90}
      >
        <option
          key="0"
          value={1}
        >
          January
        </option>
        <option
          key="1"
          value={2}
        >
          February
        </option>
        <option
          key="2"
          value={3}
        >
          March
        </option>
        <option
          key="3"
          value={4}
        >
          April
        </option>
        <option
          key="4"
          value={5}
        >
          May
        </option>
        <option
          key="5"
          value={6}
        >
          June
        </option>
        <option
          key="6"
          value={7}
        >
          July
        </option>
        <option
          key="7"
          value={8}
        >
          August
        </option>
        <option
          key="8"
          value={9}
        >
          September
        </option>
        <option
          key="9"
          value={10}
        >
          October
        </option>
        <option
          key="10"
          value={11}
        >
          November
        </option>
        <option
          key="11"
          value={12}
        >
          December
        </option>
      </withTheme(Select)>
      <withTheme(Select)
        defaultValue={2018}
        flex="unset"
        name="year"
        onChange={[Function]}
        width={70}
      >
        <option
          key="2015"
          value="2015"
        >
          2015
        </option>
        <option
          key="2016"
          value="2016"
        >
          2016
        </option>
        <option
          key="2017"
          value="2017"
        >
          2017
        </option>
        <option
          key="2018"
          value="2018"
        >
          2018
        </option>
        <option
          key="2019"
          value="2019"
        >
          2019
        </option>
        <option
          key="2020"
          value="2020"
        >
          2020
        </option>
        <option
          key="2021"
          value="2021"
        >
          2021
        </option>
      </withTheme(Select)>
    </withTheme(Pane)>
  </withTheme(Pane)>
  <withTheme(Pane)
    display="flex"
    flexDirection="row"
  >
    <withTheme(Pane)
      margin="80px"
      textAlign="center"
      width="100%"
    >
      <h2>
        Uh-oh!
      </h2>
      <p>
        It looks like you don't have a budget for this month
      </p>
    </withTheme(Pane)>
    <withTheme(Pane)
      display="flex"
      flex="1"
    />
    <withTheme(Pane)
      display="flex"
      flex="1"
    >
      <ul
        className="stat-list"
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </withTheme(Pane)>
  </withTheme(Pane)>
</withTheme(Pane)>
`;
